Fatal error “Undefined boolean filter value” when attempting to install the compiler

Recently, on the compiler trunk (I tested on https://github.com/ocaml/ocaml/commit/6a0a0a469315eeafc926f148debd52b52da2bea0), trying to install the compiler with `--assume-built` fails with a fatal error:

```
$ ./configure
$ make -j
$ opam install . --assume-built
ocaml-variants is now pinned to git+file:///home/olivier/compiler/trunk#simplify_raise (version 5.2.0+trunk)
Fatal error: Undefined boolean filter value: arch != "arm64"
Backtrace:
  Raised at Stdlib.failwith in file "stdlib.ml", line 29, characters 17-33
  Called from OpamFilter.eval_to_bool in file "src/format/opamFilter.ml" (inlined), line 378, characters 34-68
  Called from OpamFilter.filter_constraints.(fun) in file "src/format/opamFilter.ml", line 502, characters 11-40
  Called from OpamFormula.partial_eval in file "src/format/opamFormula.ml", line 263, characters 11-30
  Called from OpamFormula.partial_eval in file "src/format/opamFormula.ml", line 263, characters 11-30
  Called from OpamFormula.partial_eval in file "src/format/opamFormula.ml", line 263, characters 11-30
  Called from OpamFormula.partial_eval in file "src/format/opamFormula.ml", line 263, characters 11-30
  Called from OpamFilter.gen_filter_formula.(fun) in file "src/format/opamFilter.ml", line 536, characters 29-43
  Called from OpamFormula.map in file "src/format/opamFormula.ml", line 139, characters 35-44
  Called from OpamStd.Op.(|>) in file "src/core/opamStd.ml" (inlined), line 1385, characters 17-20
  Called from OpamClient.check_installed.(fun) in file "src/client/opamClient.ml", line 1019, characters 14-132
  Called from OpamStd.Set.Make.fold.(fun) in file "src/core/opamStd.ml", line 190, characters 15-23
  Called from Stdlib__Set.Make.iter in file "set.ml", line 378, characters 35-38
  Called from OpamStd.Set.Make.fold in file "src/core/opamStd.ml", line 189, characters 6-62
  Called from OpamClient.check_installed.(fun) in file "src/client/opamClient.ml", line 1014, characters 8-866
  Called from OpamStd.Map.Make.fold.(fun) in file "src/core/opamStd.ml", line 292, characters 14-28
  Called from Stdlib__Map.Make.iter in file "map.ml", line 297, characters 20-25
  Called from OpamStd.Map.Make.fold in file "src/core/opamStd.ml", line 291, characters 6-72
  Called from OpamClient.assume_built_restrictions in file "src/client/opamClient.ml", line 1050, characters 16-64
  Called from OpamClient.install_t in file "src/client/opamClient.ml", line 1236, characters 6-59
  Called from OpamCommands.install.install.(fun) in file "src/client/opamCommands.ml", line 1675, characters 6-157
  Called from OpamSwitchState.with_.(fun) in file "src/state/opamSwitchState.ml", line 1235, characters 14-18
  Re-raised at OpamStd.Exn.finalise in file "src/core/opamStd.ml", line 1372, characters 4-38
  Called from OpamStd.Exn.finally in file "src/core/opamStd.ml", line 1375, characters 10-14
  Re-raised at OpamStd.Exn.finalise in file "src/core/opamStd.ml", line 1372, characters 4-38
  Called from OpamGlobalState.with_ in file "src/state/opamGlobalState.ml", line 186, characters 14-18
  Re-raised at OpamStd.Exn.finalise in file "src/core/opamStd.ml", line 1372, characters 4-38
  Called from Cmdliner_term.app.(fun) in file "src_ext/cmdliner/src/cmdliner_term.ml", line 27, characters 19-24
  Called from Cmdliner.Term.ret.(fun) in file "src_ext/cmdliner/src/cmdliner.ml", line 27, characters 27-34
  Called from Cmdliner.Term.run in file "src_ext/cmdliner/src/cmdliner.ml", line 118, characters 32-39
  Called from Cmdliner.Term.term_eval in file "src_ext/cmdliner/src/cmdliner.ml", line 148, characters 18-36
  Called from Cmdliner.Term.eval_choice in file "src_ext/cmdliner/src/cmdliner.ml", line 266, characters 22-48
  Called from OpamCliMain.run.(fun) in file "src/client/opamCliMain.ml", line 441, characters 8-60
  Called from OpamCliMain.main_catch_all in file "src/client/opamCliMain.ml", line 328, characters 6-10
```

I noticed that reverting https://github.com/ocaml/ocaml/pull/12524 makes the error go away. However, that PR does not seem to be faulty.

I’m not sure how to test this:
```
$ opam pin add opam-devel --dev git+https://github.com/rjbou/opam.git#check_installed_var_resolve
[opam-devel.2.2.0~alpha2] synchronised (git+https://github.com/rjbou/opam.git#check_installed_var_resolve)
opam-devel is now pinned to git+https://github.com/rjbou/opam.git#check_installed_var_resolve (version 2.2.0~alpha3~dev)

[ERROR] Package conflict!
  * Missing dependency:
    - opam-client >= 2.2.0~alpha3~dev
    no matching version

[NOTE] Pinning command successful, but your installed packages may be out of sync.
```

--- a/master_changes.md
+++ b/master_changes.md
@@ -14,7 +14,7 @@ users)
   * Bump to 2.2.0~alpha3~dev [#5615 @rjbou]
 
 ## Global CLI
-  * Fix `OPAMVERBOSE` setting, 0 and 1 levels was inverted: eg, "no" gives level 1, and "yes" level 0 [#5686 @smorimoto]
+  * [BUG] Fix `OPAMVERBOSE` setting, 0 and 1 levels was inverted: eg, "no" gives level 1, and "yes" level 0 [#5686 @smorimoto]
 
 ## Plugins
 
@@ -53,11 +53,11 @@ users)
 ## Source
 
 ## Lint
-  * Fix extra-files handling when linting packages from repositories, see #5068 [#5639 @rjbou]
+  * [BUG] Fix extra-files handling when linting packages from repositories, see #5068 [#5639 @rjbou]
   * Allow to mark a set of warnings as errors using a new syntax -W @1..9 [#5652 @kit-ty-kate @rjbou - fixes #5651]
 
 ## Repository
-  * Fix `OPAMCURL` and `OPAMFETCH` handling [#5607 @rjbou - fix #5597]
+  * [BUG] Fix `OPAMCURL` and `OPAMFETCH` handling [#5607 @rjbou - fix #5597]
 
 ## Lock
 
@@ -78,7 +78,7 @@ users)
 ## Build
   * Remove `bigarray` dependency [#5612 @kit-ty-kate]
   * Remove use of deprecated `Printf.kprintf" [#5612 @kit-ty-kate]
-  * Fix "make cold" on Windows when gcc is available [#5635 @kit-ty-kate - fixes #5600]
+  * [BUG] Fix "make cold" on Windows when gcc is available [#5635 @kit-ty-kate - fixes #5600]
 
 ## Infrastructure
   * Test OCaml 5.0 and 5.1 in CI [#5672 @kit-ty-kate]

--- a/master_changes.md
+++ b/master_changes.md
@@ -26,6 +26,7 @@ users)
 
 ## Install
   * [BUG] On install driven by `.install` file, track intermediate directories too, in order to have them suppressed at package removal [#5691 @rjbou - fix #5688]
+  * [BUG] With `--assume-built`, resolve variables in depends filter according switch & global environment, not only depends predefined variables [#570 @rjbou - fix #5698]
 
 ## Remove
 

--- a/master_changes.md
+++ b/master_changes.md
@@ -117,6 +117,7 @@ users)
   * Add test for `OpamCoreConfig`, to check `OPAMVERBOSE` values [#5686 @rjbou]
   * dot-install: generalise inner script & use less generic filenames [#5691 @rjbou]
   * dot-install: add a test for removal of non specified in .install empty directories [#5701 @rjbou]
+  * Add test in assume-built for depends with switch variable filters [#5700 @rjbou]
 
 ### Engine
   * With real path resolved for all opam temp dir, remove `/private` from mac temp dir regexp [#5654 @rjbou]

--- a/src/client/opamClient.ml
+++ b/src/client/opamClient.ml
@@ -1078,9 +1078,14 @@ let check_installed ~build ~post t atoms =
   let test = OpamStateConfig.(!r.build_test) in
   let doc = OpamStateConfig.(!r.build_doc) in
   let dev_setup = OpamStateConfig.(!r.dev_setup) in
-  let env p =
-    OpamFilter.deps_var_env ~build ~post ~test ~doc ~dev_setup
-      ~dev:(OpamSwitchState.is_dev_package t p)
+  let env package var =
+    let content =
+      OpamFilter.deps_var_env ~build ~post ~test ~doc ~dev_setup
+        ~dev:(OpamSwitchState.is_dev_package t package) var
+    in
+    match content with
+    | None -> OpamPackageVar.resolve_switch ~package t var
+    | Some _ -> content
   in
   OpamPackage.Name.Map.fold (fun name versions map ->
       let compliant, missing_opt =

--- a/tests/reftests/assume-built.test
+++ b/tests/reftests/assume-built.test
@@ -216,7 +216,12 @@ Done.
 [NOTE] Ignoring uncommitted changes in ${BASEDIR}/ongoing (`--working-dir' not specified or specified with no argument).
 [ongoing.1.0] synchronised (no changes)
 
-Fatal error: Undefined boolean filter value: version != "1.0"
-# Return code 99 #
+The following actions will be performed:
+=== install 1 package
+  - install ongoing 1.0 (pinned)
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> retrieved ongoing.1.0  (no changes)
+-> installed ongoing.1.0
+Done.
 ### test -f ./ongoing/installed
-# Return code 1 #

--- a/tests/reftests/assume-built.test
+++ b/tests/reftests/assume-built.test
@@ -166,3 +166,57 @@ Processing  2/2: [ongoing: touch installed]
 -> installed ongoing.dev
 Done.
 ### test -f ./ongoing/installed
+### : assume built & opam file variables :
+### opam unpin ongoing
+Ok, ongoing is no longer pinned to git+file://${BASEDIR}/ongoing#master (version dev)
+The following actions will be performed:
+=== remove 1 package
+  - remove ongoing dev
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> removed   ongoing.dev
+Done.
+### opam var os=some-os --global
+Added '[os "some-os" "Set through 'opam var'"]' to field global-variables in global configuration
+### <pin:ongoing/ongoing.opam>
+opam-version: "2.0"
+version: "1.0"
+depends: [
+  "qux" { with-test }
+  "foo" { os != "some-os" }
+  "bar" { version != "1.0" }
+]
+build: [
+  [ "test" "-f" "ongoing.txt" ]
+  [ "sh" "-c" "cat ongoing.txt > out" ]
+]
+install: [
+  ["test" "-f" "out" ]
+  [ "test" "-f" "ongoing.txt" ] { with-test }
+  [ "test" "-f" "out2" ] { os != "some-os" }
+  [ "test" "-f" "out3" ] { version != "1.0" }
+  ["touch" "installed"]
+]
+### git -C ongoing clean -fdqx
+### opam pin ./ongoing -n
+Package ongoing does not exist, create as a NEW package? [y/n] y
+ongoing is now pinned to git+file://${BASEDIR}/ongoing#master (version 1.0)
+### touch ongoing/out
+### opam install qux
+The following actions will be performed:
+=== install 1 package
+  - install qux u
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> installed qux.u
+Done.
+### opam install ongoing --assume-built --with-test
+
+<><> Synchronising pinned packages ><><><><><><><><><><><><><><><><><><><><><><>
+[NOTE] Ignoring uncommitted changes in ${BASEDIR}/ongoing (`--working-dir' not specified or specified with no argument).
+[ongoing.1.0] synchronised (no changes)
+
+Fatal error: Undefined boolean filter value: version != "1.0"
+# Return code 99 #
+### test -f ./ongoing/installed
+# Return code 1 #

